eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/ChannelRelayer.scala

```diff
@@ -16,220 +16,107 @@
-            val channelInfo_opt = channelUpdates.get(shortChannelId)
-            val channelUpdate_opt = channelInfo_opt.map(_.channelUpdate)
-            val relayResult = relayOrFail(replyTo, relayPacket, channelUpdate_opt)
-            log.debug(s"candidate channel for htlc #${add.id}: shortChannelId={} balanceMsat={} channelUpdate={} relayResult={}", shortChannelId, channelInfo_opt.map(_.commitments.availableBalanceForSend).getOrElse(""), channelUpdate_opt.getOrElse(""), relayResult)
-            (shortChannelId, channelInfo_opt, relayResult)
-          }
-          .collect { case (shortChannelId, Some(channelInfo), _: RelaySuccess) => (shortChannelId, channelInfo.commitments.availableBalanceForSend) }
-          .filter(_._2 > relayPacket.payload.amountToForward) // we only keep channels that have enough balance to handle this payment
-          .toList // needed for ordering
-          .sortBy(_._2) // we want to use the channel with the lowest available balance that can process the payment
-          .headOption match {
-          case Some((preferredShortChannelId, availableBalanceMsat)) if preferredShortChannelId != requestedShortChannelId =>
-            log.info("replacing requestedShortChannelId={} by preferredShortChannelId={} with availableBalanceMsat={}", requestedShortChannelId, preferredShortChannelId, availableBalanceMsat)
-            Some(preferredShortChannelId)
-          case Some(_) =>
-            // the requested short_channel_id is already our preferred channel
-            Some(requestedShortChannelId)
-          case None if !alreadyTried.contains(requestedShortChannelId) =>
-            // no channel seem to work for this payment, we keep the requested channel id
-            Some(requestedShortChannelId)
-          case None =>
-            // no channel seem to work for this payment and we have already tried the requested channel id: we give up
-            None
+  def apply(
+             nodeParams: NodeParams,
+             register: ActorRef,
+             channelUpdates: ChannelUpdates = Map.empty,
+             node2channels: NodeChannels = mutable.MultiDict.empty[PublicKey, ShortChannelId]
+           ): Behavior[Command] =
+    Behaviors.setup { context =>
+      context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[LocalChannelUpdate](WrappedLocalChannelUpdate))
+      context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[LocalChannelDown](WrappedLocalChannelDown))
+      context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[AvailableBalanceChanged](WrappedAvailableBalanceChanged))
+      context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[ShortChannelIdAssigned](WrappedShortChannelIdAssigned))
+      context.messageAdapter[IncomingPacket.ChannelRelayPacket](WrappedChannelRelayPacket)
+      Behaviors.withMdc(Logs.mdc(category_opt = Some(Logs.LogCategory.PAYMENT)), mdc) {
+        Behaviors.receiveMessagePartial {
```

Comment:
To be honest, I used `receiveMessage`/`receiveMessagePartial` interchangeably without noticing the difference. `receiveMessage` seems stricter but now I'm getting those warnings:
```
Exhaustivity analysis reached max recursion depth, not all missing cases are reported.
(Please try with scalac -Ypatmat-exhaust-depth 40 or -Ypatmat-exhaust-depth off.)
    Behaviors.receiveMessage {
```

But I can't get this configuration to be accepted by `scala-maven-plugin` right away. Needs more digging.

Reply:
Ok, found out what the difference is.

With `receiveMessage`, any unhandled message will cause the actor to fail (that is, to stop), whereas with `receiveMessagePartial` it will just print a deadletter log. That's why the former generates a compilation warning. Obviously we should use `receiveMessage` as much as possible.

Reply:
Actually, if we model our typed actors as FSM with behaviors as states, we have no other choise than using `receiveMessagePartial` almost everywhere, since each state typically handles a subset of all the channel commands. We could use `receiveMessage` and manually ignore all unmatched message, but that's exactly what `receiveMessagePartial` does.

Side note : there are limits to what the compiler can detect apparently. E.g. in the `sending` method below, not all messages are handled but there are no warning (probably due to the `orElse` construct).

https://github.com/ACINQ/eclair/blob/typed-relayers/eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/NodeRelay.scala#L236-L265

Reply:
In that case, do you think we could have a top-level `sealed trait Command` and below that, other `sealed trait SomeStateCommand extends Command` for each of the FSM states, and in each state declare ourselves as `ActorRef[SomeStateCommand]`? Or is it overkill?

Reply:
We could, but then we'll have to use `Behaviors.[SubCommand].unsafeUpcast[Command]` somewhere. That may be better because more explicit, I'm not sure.

Reply:
Would it be better with the less functional way of implementing the actor?

Reply:
It could even get messier if we have three states: 

```scala

sealed trait Command
sealed trait State1Command extends Command
sealed trait State2Command extends Command

def apply(): Behavior[Command] = state1.unsafeUpcast[Command]

def state1(): Behavior[State1Command] = Behaviors.receiveMessage {
  case s1: State1Command => state2.unsafeUpcast[State1Command]
}

def state2(): Behavior[State2Command] = Behaviors.receiveMessage {
  case s2: State2Command => Behaviors.stopped
}
```

The `state2.unsafeUpcast[State1Command].unsafeUpcast[Command]` is... let's say not very elegant.

Reply:
> Would it be better with the less functional way of implementing the actor?

I don't think so, this is a logical problem, not a technical one. We could also do:


```scala

sealed trait Command
sealed trait State1Command extends Command
sealed trait State2Command extends Command

def apply(): Behavior[Command] = state1.unsafeUpcast[Command]

def state1(): Behavior[Command] = Behaviors.receiveMessage[State1Command] {
  ...
}.unsafeUpcast[State1Command]

def state2(): Behavior[Command] = Behaviors.receiveMessage[State2Command] {
  ...
}.unsafeUpcast[State2Command]
```

Reply:
It's weird, what we're doing really sounds like a very common usage of akka typed, it's surprising that there isn't a more elegant solution :confused:

Reply:
In the [akka doc](https://doc.akka.io/docs/akka/current/typed/fsm.html#behaviors-as-finite-state-machines) they use a combination of `receiveMessage` + `case _ => Behaviors.unhandled` for one state, and `receivePartial` for another one. Looks like there isn't a real good way of doing what we want to do.